[FIX] project: portal visibility of a task

yosa-odoo opened the pull request:
Steps to reproduce:
- Install Project
- Create internal user with usable adress mail
- Create portal user
- Define the mail settings for ingoing/outgoing server
- Create a new project with an alias for the mails to be fetched; in settings put the portal user as visible
- Send an email with the email adress of the internal user
-> In the task created, the portal user does not appear

OPW-2704113
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/82160).

yosa-odoo commented:
I wanted to change the variables from which the method is dependent: 
https://github.com/odoo/odoo/blob/e6d77a43e9c15e6cb4969a50bbf0f5b10273c980/addons/project/models/project.py#L875-L876

by adding `project_id.task_ids` (or other fields related to a change in the project in the creation of a new task by email) to recompute the allowed users when a new task is added to a project. But, the compute is triggered with a bad timing. The only way (I found) is to call it manually just after adding followers to the task.

IT-Ideas commented on addons/project/models/project.py:
```diff
@@ -1305,6 +1305,7 @@ def message_new(self, msg, custom_values=None):
         customer_ids = [p.id for p in self.env['mail.thread']._mail_find_partner_from_emails(tools.email_split(defaults['email_from']), records=task) if p]
         partner_ids += customer_ids
         task.message_subscribe(partner_ids)
+        task._compute_allowed_user_ids()
```
I think that the explanation is right [here](https://github.com/odoo/odoo/blob/6b084c7ace17260eb3e41f2840190a36e75eb358/addons/project/models/project.py#L883-L884)
By calling the compute here, which solve the issue, you actually run the code highlighted in the link.
I think it would be best to take the logic and apply it here directly to the partner_ids prior to call `task.message_subscribe(partner_ids)` instead of calling the compute.
What do you think?
If I remember correctly the right system has been changed somewhere between 14.0 and 15.0. Can you check? If it is the case, can you check that the issue is still happening in 15.0? If not, let me know so that we prevent the forward bot to run.

yosa-odoo replied:
In V15, the field visible is not there anymore. The way it works is adding the customer/portal to the followers. So, yes we should prevent the forward port.

yosa-odoo replied:
```I think it would be best to take the logic and apply it here directly to the partner_ids prior to call task.message_subscribe(partner_ids) instead of calling the compute.```
I'll try this approach

IT-Ideas reviewed (changes requested):
Thanks for the fix 🙂
I would do it slightly differently.
Could you also add a test something similar to [here](https://github.com/odoo/odoo/blob/1940c2371e64b06a6df7a412edf203d7f81838ed/addons/account/tests/test_account_incoming_supplier_invoice.py#L52) but adapted to your use case of course 🙂.
Thanks for you researches and work 🙂

nle-odoo reviewed (approved):
looks good to me, the code itself says to add the partner to current allowed users (and if there is no partner to add, it is working as expecting and adding the project allowed users)

I think you could also do:

```diff
diff --git a/addons/project/models/project.py b/addons/project/models/project.py
index 16cfe696a7d..8d5c132aa18 100644
--- a/addons/project/models/project.py
+++ b/addons/project/models/project.py
@@ -1006,6 +1006,7 @@ class Task(models.Model):
         if partner_ids:
             new_allowed_users = self.env['res.partner'].browse(partner_ids).user_ids.filtered('share')
             tasks = self.filtered(lambda task: task.project_id.privacy_visibility == 'portal')
+            tasks.allowed_user_ids # ensure field is computed before modifying it
             tasks.sudo().write({'allowed_user_ids': [(4, user.id) for user in new_allowed_users]})
         return res
```

so the field is computed normally instead of calling the compute method. but in this case both code are equivalent I think.

it would probably be good to have a test in

addons/project/tests/test_project_flow.py

that ensures the use case is working (project privacy_visibility='portal', mail is sent by existing user => test that the project allowed users is no longer lost)

yosa-odoo commented:
@nle-odoo I changed the code. I think your way is more intuitive in the understanding of the code flow.

I added the tests. The test fails without the fix and works with fix.

IT-Ideas commented on addons/project/models/project.py:
```diff
@@ -1006,6 +1006,7 @@ def message_subscribe(self, partner_ids=None, channel_ids=None, subtype_ids=None
         if partner_ids:
             new_allowed_users = self.env['res.partner'].browse(partner_ids).user_ids.filtered('share')
             tasks = self.filtered(lambda task: task.project_id.privacy_visibility == 'portal')
+            tasks.allowed_user_ids  # ensure field is computed before modifying it
             tasks.sudo().write({'allowed_user_ids': [(4, user.id) for user in new_allowed_users]})
```
What about this?
I find this more elegant than calling the field value before the write.
I might be wrong but I think it will do the exact same thing. What do you think @nle-odoo ?
Thanks for your help guys 🙂 
```suggestion
            tasks.sudo().allowed_user_ids |= user
```

yosa-odoo replied:
```suggestion
            tasks.allowed_user_ids  # ensure field is computed before modifying it
            tasks.sudo().write({'allowed_user_ids': [(4, user.id) for user in new_allowed_users]})
            tasks.sudo().allowed_user_ids |= new_allowed_users
```

yosa-odoo replied:
It passes the test

IT-Ideas replied:
> So it would be?:

Indeed, sorry!

nle-odoo replied:
looks good

IT-Ideas commented:
@fw-bot ignore

IT-Ideas commented:
@robodoo r+
